### ISSUE TITLE
Add support for array data type handling

### ISSUE THREAD

**Issue by JulienBedel** [Orange-Cyberdefense/bof#13]
### TL;DR

OPC UA sometimes receives arrays of strings.

Array data handling has not been thought and purposely implemented in BOF yet. However workarounds might exists to support array with BOF current functionalities.

### Description

**Context**

I encountered this problem while implementing OPC UA protocol into BOF.

**Problem**

OPC UA sometimes receives arrays of strings.
Array data handling has not been thought and purposely implemented in BOF yet.

In fact, BOF has been developed with the idea to work with a fixed number of block/fields. In other words, even if some items are optional at least we know that they are here.

### Solutions to investigate

I found it hard to think of a proper solution using BOF current implementation.
Here are some ideas I have but that have no example associated. I will still try to put it into words :

**1. Make array elements optional**

Even if arrays have variable size, they have a limited number of elements they can contain. Therefore we might define every element of an array as optional and make them present depending on the specified array size.

The issue is then "how to make the link between the array length and the optional elements ?".

**2. Recursivity**

Add a recursive mechanism in BOF to allow an array element to refer to the next one (as in a chained list) until we reached the end. How could this be implemented in BOF ?

**3. Making BOF include types**

This is quite an overkill solution but might solve other difficulties : maybe implementing a type mechanism into BOF that transform specific types into bytes according to a protocol specification would be a way to use arrays ?

In fact, for now BOF is handling fields as raw bytes which size depends on protocol specification, but does not take any typing into account. We could define a specification in JSON file using types, and make the type association in a protocol-specific JSON too ?

### Help needed

Any help is appreciated, more precisely regarding the following topics :

- Clarify the problem and solution ideas.
-  Identify the appropriate solution in proposed one, or a new one if it fits the problem.
- Help fixing the issue.

**Comment by claire-lex** [comment]
This issue is not relevant anymore, as our next release relies on Scapy for protocol implementation, which supports (enforces?) types (Preview in branch `scapybof`).
The implementation of OPC UA is missing from Scapy, though.
